### PR TITLE
A few changes to federated-service e2e test.

### PR DESCRIPTION
**Pull request by mml**
Most of the changes that get the test to pass have been made already or
elsewhere.  Here we restructure a bit fixing a nesting problem, extend the
timeouts, and start creating distinct backend pods that I'll delete in the
non-local test (coming shortly).

Also some extra debugging info in the DNS code.  I made some upstream
changes to skydns in https://github.com/skynetservices/skydns/pull/283

For #27739

Includes a commit from @madhusudancs that I will remove once his merges.

### CODE REVIEW

**Comment by nikhiljindal** [comment]
The DNS changes here are just code readability changes and no functional change, right @mml ?

**Comment by mml** [comment]
and debug logs, but no functionality.

File: test/e2e/federated-service.go [hunk 1]
```diff
@@ -223,28 +233,23 @@ func waitForFederatedServiceShard(cs *release_1_3.Clientset, namespace string, s
...
-		clSvc.Spec.ClusterIP = ""
-
 		Expect(clSvc.Name).To(Equal(service.Name))
 		// Some fields are expected to be different, so make them the same before checking equality.
 		clSvc.Spec.ClusterIP = service.Spec.ClusterIP
 		clSvc.Spec.ExternalIPs = service.Spec.ExternalIPs
 		clSvc.Spec.DeprecatedPublicIPs = service.Spec.DeprecatedPublicIPs
 		clSvc.Spec.LoadBalancerIP = service.Spec.LoadBalancerIP
 		clSvc.Spec.LoadBalancerSourceRanges = service.Spec.LoadBalancerSourceRanges
+		for i := range clSvc.Spec.Ports {
```
Comment (quinton-hoole) [inline test/e2e/federated-service.go]: minor nit: Prefer "for i, port := ....".  Then you can just use "port.NodePort below.

File: test/e2e/federated-service.go [hunk 2]
```diff
@@ -344,10 +349,36 @@ func discoverService(f *framework.Framework, name string, exists bool) {
 
 	if exists {
 		// TODO(mml): Eventually check the IP address is correct, too.
-		Eventually(podExitCodeDetector(f, FederatedServicePod, 0), DNSTTL, time.Second*2).
+		Eventually(podExitCodeDetector(f, FederatedServicePod, 0), 10*DNSTTL, time.Second*2).
```
Comment (quinton-hoole) [inline test/e2e/federated-service.go]: 10 times seems excessive?  Surely this should succeed within 2 * DNSTTL?

File: test/e2e/federated-service.go [hunk 3]
```diff
@@ -344,10 +349,36 @@ func discoverService(f *framework.Framework, name string, exists bool) {
...
 			Should(BeNil(), "%q should exit 0, but it never did", command)
 	} else {
-		Consistently(podExitCodeDetector(f, FederatedServicePod, 0), DNSTTL, time.Second*2).
+		Consistently(podExitCodeDetector(f, FederatedServicePod, 0), 10*DNSTTL, time.Second*2).
```
Comment (quinton-hoole) [inline test/e2e/federated-service.go]: As above.

File: test/e2e/federated-service.go [hunk 4]
```diff
@@ -344,10 +349,36 @@ func discoverService(f *framework.Framework, name string, exists bool) {
...
 			ShouldNot(BeNil(), "%q should never exit 0, but it did", command)
 	}
 }
+
+func createBackendPods(clusterClientSets []*release_1_3.Clientset, namespace string) {
+	name := "backend"
+
+	pod := &v1.Pod{
```
Comment (quinton-hoole) [inline test/e2e/federated-service.go]: nit: Probably more resilient if you create a replcaset of 1 here.  That way if it crashes or ends up on a dead node or similar, a new one will be created automatically.

**Comment by quinton-hoole** [comment]
Thanks @mml.  Only nits, which can be addressed in a followup PR if you prefer.

**Comment by quinton-hoole** [comment]
PS:  Feel free to remove the LGTM label to avoid auto-merge if you prefer to address the nits in this PR.

File: test/e2e/federated-service.go (hunk 1 above)
Comment (mml) [inline test/e2e/federated-service.go]: That actually doesn't work, as port is a value type and ends up being a copy of the object and your changes get lost.

I'll add a comment explaining this, as I banged my head against it for quite awhile.

Comment (quinton-hoole) [inline test/e2e/federated-service.go]: Doh!  Of course you're right :-)

**Comment by mml** [comment]
I managed to drop some changes during my git antics, but I think it is in good shape now.

File: pkg/dns/dns.go [hunk 5]
```diff
@@ -559,22 +559,38 @@ func getSkyMsg(ip string, port int) (*skymsg.Service, string) {
...
-		for i, domComp := range kd.domainPath {
-			// kd.domainPath is reversed, so we need to look in the `path` in the reverse order.
-			if domComp != path[len(path)-i-1] {
-				return false
-			}
-		}
-		if _, ok := kd.federations[path[2]]; ok {
-			return true
+	if len(path) != 4+len(kd.domainPath) {
+		glog.Infof("len(%q) != 4+len(%q)", path, kd.domainPath)
```
Comment (nikhiljindal) [inline pkg/dns/dns.go]: It might be better to move these glog.Infos to glog.V(2).Info.

File: pkg/dns/dns.go [hunk 6]
```diff
@@ -559,22 +559,38 @@ func getSkyMsg(ip string, port int) (*skymsg.Service, string) {
...
+		return false
+	}
+	if errs := validation.IsDNS952Label(path[0]); len(errs) != 0 {
+		glog.Infof("%q is not an RFC 952 label: %q", path[0], errs)
+		return false
+	}
+	if errs := validation.IsDNS1123Label(path[1]); len(errs) != 0 {
+		glog.Infof("%q is not an RFC 1123 label: %q", path[1], errs)
```
Comment (nikhiljindal) [inline pkg/dns/dns.go]: Would a better log message be: "not a federation query as %q is not an RFC 1123 label: %q"

**Comment by quinton-hoole** [comment]
"Jenkins GKE Smoke Test" is something I've not seen before.  It seems to be borked.  @lavalamp ?

@k8s-bot test this issue: #IGNORE

```
+ gsutil -mq cp -r gs://cloud-sdk-testing/ci/staging /var/lib/jenkins/workspace/kubernetes-pull-build-test-e2e-gke
AccessDeniedException: 403 Forbidden
CommandException: 1 file/object could not be transferred.
+ sleep 1
+ rm -rf /var/lib/jenkins/workspace/kubernetes-pull-build-test-e2e-gke/.gsutil
+ for n in '$(seq 3)'
+ gsutil -mq cp -r gs://cloud-sdk-testing/ci/staging /var/lib/jenkins/workspace/kubernetes-pull-build-test-e2e-gke
AccessDeniedException: 403 Forbidden
CommandException: 1 file/object could not be transferred.
+ sleep 1
+ rm -rf /var/lib/jenkins/workspace/kubernetes-pull-build-test-e2e-gke/.gsutil
+ for n in '$(seq 3)'
+ gsutil -mq cp -r gs://cloud-sdk-testing/ci/staging /var/lib/jenkins/workspace/kubernetes-pull-build-test-e2e-gke
AccessDeniedException: 403 Forbidden
CommandException: 1 file/object could not be transferred.
+ sleep 1
+ rm -rf /var/lib/jenkins/workspace/kubernetes-pull-build-test-e2e-gke/.gsutil
+ rm -rf /var/lib/jenkins/workspace/kubernetes-pull-build-test-e2e-gke/repo /var/lib/jenkins/workspace/kubernetes-pull-build-test-e2e-gke/cloudsdk
++ basename gs://cloud-sdk-testing/ci/staging
+ mv /var/lib/jenkins/workspace/kubernetes-pull-build-test-e2e-gke/staging /var/lib/jenkins/workspace/kubernetes-pull-build-test-e2e-gke/repo
mv: cannot stat `/var/lib/jenkins/workspace/kubernetes-pull-build-test-e2e-gke/staging': No such file or directory
+ rc=1
+ [[ 1 -ne 0 ]]
+ [[ -x cluster/log-dump.sh ]]
+ [[ -d _artifacts ]]
+ [[ 1 -eq 124 ]]
+ [[ 1 -eq 137 ]]
+ [[ 1 -ne 0 ]]
+ echo 'Build failed'
Build failed
+ echo 'Exiting with code: 1'
Exiting with code: 1
+ exit 1
```

**Pull request merged**